// @fawry_pay/rn-fawry-pay-registered-sdk/src/index.tsx
import { NativeEventEmitter, NativeModules, Platform } from 'react-native';

const LINKING_ERROR = `
The package '@fawry_pay/rn-fawry-pay-sdk' doesn't seem to be linked. Make sure:

${Platform.select({ ios: "- You have run 'pod install'\n", default: '' })}
- You rebuilt the app after installing the package
- You are not using Expo Go
`;

const RnFawryPaySdk =
  NativeModules.RnFawryPaySdk ||
  new Proxy(
    {},
    {
      get() {
        throw new Error(LINKING_ERROR);
      },
    }
  );

export async function startPayment(
  addressHierarchy: AddressHierarchy,
  allow3DPayment: boolean,
  apiPath: string,
  baseUrl: string,
  customerInfo: CustomerInfo,
  items: BillItems[],
  lang: FawryLanguages,
  merchantInfo: MerchantInfo,
  showLoyaltyContainer: boolean,
  showTipsView: boolean,
  showVoucherContainer: boolean,
  skipReceipt: boolean,
  beid?: string,
  branchCode?: string,
  branchName?: string,
  scheduledTime?: string,
  serviceTypeCode?: string,
  shippingAddress?: Address,
  tableId?: number
): Promise<any> {
  
  const optionalParams: OptionalPaymentParams = {
    beid,
    branchCode,
    branchName,
    scheduledTime,
    serviceTypeCode,
    shippingAddress,
    tableId,
  };

  RnFawryPaySdk.startPayment(
    baseUrl,
    lang,
    addressHierarchy,
    allow3DPayment,
    apiPath,
    customerInfo,
    items,
    merchantInfo,
    showLoyaltyContainer,
    showTipsView,
    showVoucherContainer,
    skipReceipt,
    optionalParams
  );
}

export async function openCardsManager(
  baseUrl: string,
  lang: FawryLanguages,
  merchantInfo: MerchantInfo,
  customerInfo: CustomerInfo
): Promise<any> {
  RnFawryPaySdk.openCardsManager(baseUrl, lang, merchantInfo, customerInfo);
}

export async function openAddressManager(
  baseUrl: string,
  lang: FawryLanguages,
  merchantInfo: MerchantInfo,
  customerInfo: CustomerInfo,
  beid: string,
  addressHierarchy: AddressHierarchy
): Promise<any> {
  RnFawryPaySdk.openAddressManager(
    baseUrl,
    lang,
    merchantInfo,
    customerInfo,
    beid,
    addressHierarchy
  );
}

export interface BillItems {
  itemId: string;
  description: string;
  quantity: string;
  price: string;
  originalPrice: string;
  width?: string;
  height?: string;
  weight?: string;
  variantCode?: string;
  earningRuleId?: string;
  imageUrl?: string;
  specialRequest?: string;
  tax: number;
}

export interface MerchantInfo {
  merchantCode: string;
  subMerchantCode?: string;
  merchantSecretCode: string;
  merchantRefNum: string;
}

export interface CustomerInfo {
  customerName: string;
  customerMobile: string;
  customerEmail: string;
  customerProfileId: string;
  customerCif: string;
  customerToken: string;
}

export interface FawryLaunchModel {
  addressHierarchy: AddressHierarchy;
  allow3DPayment: boolean;
  apiPath: string;
  baseUrl: string;
  beid: string;
  branchCode: string;
  branchName: string;
  customerInfo: CustomerInfo;
  items: BillItems[];
  lang: FawryLanguages;
  merchantInfo: MerchantInfo;
  scheduledTime: string;
  serviceTypeCode: string;
  shippingAddress: Address;
  showLoyaltyContainer: boolean;
  showTipsView: boolean;
  showVoucherContainer: boolean;
  skipReceipt: boolean;
  tableId: number;
}

export enum FawryLanguages {
  ENGLISH = 'ENGLISH',
  ARABIC = 'ARABIC',
}

export enum AddressHierarchy {
  MATRIX = 'MATRIX',
  GEOLOCATION = 'GEOLOCATION',
}

export class Address {
  status?: string;
  apartmentNumber?: string;
  id?: number;
  addressType?: AddressType;
  default?: boolean;
  isDefault?: boolean;
  area?: AddressType;
  governorate?: AddressType;
  address?: string;
  receiverName?: string;
  city?: AddressType;
  buildingNumber?: string;
  receiverMobile?: string;
  floorNumber?: number;
  geolocation?: Geolocation;
  selected?: boolean;
}

export class AddressType {
  namePrimaryLang?: string;
  id?: string;
  nameSecondaryLang?: string;
  code?: string;
}

export class Geolocation {
  areaName?: string;
  cityName?: string;
  govName?: string;
  id?: number;
  latitude?: number;
  longitude?: number;
  streetName?: string;
}

class OptionalPaymentParams {
  beid?: string;
  branchCode?: string;
  branchName?: string;
  scheduledTime?: string;
  serviceTypeCode?: string;
  shippingAddress?: Address;
  tableId?: number;
}

export class FawryCallbacks {
  static FawryEmitter = new NativeEventEmitter(RnFawryPaySdk);
  static FAWRY_EVENT_PAYMENT_COMPLETED = 'FAWRY_EVENT_PAYMENT_COMPLETED';
  static FAWRY_EVENT_ON_SUCCESS = 'FAWRY_EVENT_ON_SUCCESS';
  static FAWRY_EVENT_ON_FAIL = 'FAWRY_EVENT_ON_FAIL';
  static FAWRY_EVENT_CARD_MANAGER_FAIL = 'FAWRY_EVENT_CARD_MANAGER_FAIL';
  static FAWRY_EVENT_ADDRESS_MANAGER_FAIL = 'FAWRY_EVENT_ADDRESS_MANAGER_FAIL';
}
